woocommerce/pinterest-for-woocommerce
# Feature/improve feedstate logic

Closes #178 
Closes #166 

### Changes proposed in this pull request

Small-ish refactor of the `ProductSync` methods related to handling the status of product feed.
More specifically;

- Improves logic of how we set & retrieve the status of the product feed by refactoring `ProductSync::feed_job_status()` onto smaller get/set methods. 
- Simplifies/reduces stored data for feed's current status.
- Uses transients for saving the related data.
- Avoids storing progress as string in the database.  
- Moves deregistration methods to concurrent request
- Avoid unnecessary calls to reschedule.
- Re-implements cleaner `reschedule_if_expired()`
- Implements `reschedule_if_errored()` to restart job after a delay.
- Remove scheduled actions on plugin deactivation. 

#### Screenshots

No visible changes to user flows / interactions. 

### Detailed test instructions

1. Start with a clean plugin installation and a new Pinterest account.
2. Start the onboarding process and Connect to Pinterest.
3. Complete onboarding process
4. Verify that the product feed is being generated properly by visiting the Catalog tab of the plugin.
5. Verify that the product feed is being registered with Pinterest properly using Pinterest's dashboard.
5. Verify that the product feed is being regenerated after it expires (24h)
6. Verify that progress is not stored as translated strings as stated in #178
6. Verify that when Product Sync is disabled, the feed file is being deleted, the related transients are being removed and the scheduled actions are being removed. 
7. Verify that when the plugin is deactivated, scheduled tasks are removed.
